### PR TITLE
Consolidate overall Global Styles mechanics in the server

### PR DESCRIPTION
**Pull request by oandregal**
This PR seeks to consolidate the overall server-side mechanism for Global Styles. Continues work from https://github.com/WordPress/gutenberg/pull/19883

1) It adds the user's Global Styles to the resolver, so it now works using core, theme, and user data. It stores the user's data using a Custom Post Type, so we can leverage existing editor mechanisms (undo/redo, draft/publish status for Global Styles, etc). This also presumes we want Global Styles to be tied to the theme so when theme changes we want to start with a new blank slate for Global Styles, which is implemented by using the theme's textdomain as part of the CPT's post name.

2) It exposes to the editor a couple of variables necessary for live-editing the styles: the ID of the entity that holds the user's styles + the base styles (core+theme). How this works will be demonstrated in subsequent PR. Note this is made to work only with the Beta Site Editor.

3) As per [past conversations](https://github.com/WordPress/gutenberg/pull/19883), it flattens the Global Styles tree stripping out the block variables. It looks like this:

```js
{
	'color': {
		'primary': '<primary color>',
		'background': '<background color>',
		'text': '<text color>',
	},
        'line-height': '<line height>'
}
```

which will be converted to this CSS rule (see [discussion about syntax](https://github.com/WordPress/gutenberg/pull/20047#discussion_r377547790)):

```js
:root {
	--wp--color--primary: '<primary color>';
	--wp--color--background: '<background color>';
	--wp--color--text: '<text color>';
        --wp--line-height: '<line height>';
}
```

Note that the names and variables we expose are still in flux, so take the current ones as only valid for demo purposes.

## How to test

- Install and enable one of the block-based themes (find one at [theme-experiments](https://github.com/WordPress/theme-experiments)). Instructions here assume Twenty Twenty Blocks.
- Check out this PR and `npm install && npm run build`.
- Create a **void** file named `experimental-theme.json` in your current theme's folder. For example, `wp-content/themes/twentytwenty-blocks/experimental-theme.json`.
- Got to Gutenberg > Experiments and enable "Full Site Editing".
- Go to Gutenberg > Site Editor.

**1) Check that core's Global Styles work:**

- Add a button and verify that its background color is `#52accc`.

**2) Check that theme's Global Styles override core's if present:**

- Add the following into the `experimental-theme.json` file created before.

```json
{
	"color": {
		"primary": "hotpink"
	}
}
```

- Reload the Site Editor.
- Add a button and verify that the button's color is now `hotpink`.

**3) Check that user's Global Styles override theme's if present:**

- Insert in the database a new post with the following data [1]:
  - post_type='wp_global_styles'
  - post_name='wp-global-styles-THEME-TEXTDOMAIN' (ex: 'wp-global-styles-twentytwenty-blocks')
  - post_content='{"color":{"primary":"blue"}}'
  - post_status='publish'
- Reload the Site Editor.
- Add a button and verify that the button's color is now `blue`.

---

[1] Make sure there isn't another post with the same post_name; if there is one, delete it first. Find below some ways you can add the required data for testing purposes. This is for the experimental Twenty Twenty Blocks theme, change to use your active theme's text-domain as required.

**wp-cli** -- you may need to provide other connection parameters depending on your setup:

```sh
wp post delete $(wp post list --post_type='wp_global_styles' --format=ids) --force
wp post create
    --post_type='wp_global_styles'
    --post_name='wp-global-styles-twentytwenty-blocks'
    --post_content='{"color":{"primary":"blue"}}'
    --post_status='publish'
```

**PHP** -- paste the following code within the `gutenberg_experimental_global_styles_enqueue_assets` when it comes to test user's Global Styles:

```php
	// Delete existing CPT, if any.
	$gs_cpt_list = wp_get_recent_posts( [
		'post_type' => 'wp_global_styles',
	] );
	foreach ( $gs_cpt_list as $gs_cpt ) {
		wp_delete_post( $gs_cpt['ID'] );
	}

	// Insert new one with required data.
	wp_insert_post( [
	    'ID'           => 0,
	    'post_content' => json_encode( [
	        'color'=> [
	            'primary' => 'blue',
	        ],
	    ] ),
	    'post_type'   => 'wp_global_styles',
	    'post_name'   => 'wp-global-styles-twentytwenty-blocks',
	    'post_status' => 'publish',
	] );
```

### CODE REVIEW

**Comment by nosolosw** [comment]
I just realized we didn't hide the GlobalStyles code behind a experimental flag. I'll add that in this PR.

**Comment by ItsJonQ** [comment]
Heya @nosolosw !! Thank you for working on this 😍 

I walked through the steps you outlined:

* [x] Check out this PR and npm install && npm run build.
* [x] Create a void file named experimental-theme.json in your current theme's folder.
* [x] Enable the "Global Styles" in the experiments page.
* [x] Create a new post with a block button, publish, and go to the front end.

However, I wasn't able to see the Global styles stuff render. No `wp-gs` className being added, no `hotpink` button.

Is there something I need to do in addition to this?

Note: Testing locally on fresh Docker install and TwentyTwenty theme.

Thanks!

**Comment by nosolosw** [comment]
Nope, that's all you need! Just tested with a brand new site, fresh clone PR, and the steps above: it worked fine for me.

If there is no `wp-gs` or `#global-styles-inline-css` stylesheet it means GlobalStyles isn't active: either the `experimental-theme.json` couldn't be found in the top-level dir of the theme or the experiment flag isn't enabled.

As per the CLI error. It looks like you need to provide some extra info (user/pass)? If you use the env bundled with gutenberg it may be more convenient to do `npm run env cli` instead. An alternative to CLI if that gets cumbersome would be pasting the following code within the `gutenberg_experimental_global_styles_enqueue_assets` function for testing user's Global Styles:

```php
	// Delete existing CPT, if any.
	$gs_cpt_list = wp_get_recent_posts( [
		'post_type' => 'wp_global_styles',
	] );
	foreach ( $gs_cpt_list as $gs_cpt ) {
		wp_delete_post( $gs_cpt['ID'] );
	}

	// Insert new one with required data.
	wp_insert_post( [
	    'ID'           => 0,
	    'post_content' => json_encode( [
	        'color'=> [
	            'primary' => 'blue',
	        ],
	    ] ),
	    'post_type'   => 'wp_global_styles',
	    'post_name'   => 'wp-global-styles-twenty-twenty', // Note this is for TwentyTwenty!
	    'post_status' => 'publish',
	] );
```

**Comment by nosolosw** [comment]
I saw that we were using the Full Site Editing flag in https://github.com/WordPress/gutenberg/pull/20061 so I thought it'd be best to use it here as well. Changed in https://github.com/WordPress/gutenberg/pull/20047/commits/f9d90e036a1fe4ee65371c92f70258bce9343a71 and updated the instructions accordingly.

**Comment by ItsJonQ** [comment]
@nosolosw Hallooo!! Thank you for your support 🙏 
Unfortunately, I'm still not able to get it working on my end

I've recorded a video here walking through the instructions:

https://www.loom.com/share/0635e0c69b1042129e09a12f555920e3

Hopefully this helps with some debugging?

**Comment by ItsJonQ** [comment]
@nosolosw Hallo!!!

Just tested this again, and it's working as expected! I have no idea why it wasn't working for me previously. Thank you so much for your efforts in helping me debug! 🤗 

It was most likely my local Docker environment. My machine has been having issues lately.
I reinstalled all Docker things 2 days ago, and it's been pretty stable since.

I left a comment for something to fix. After that, I would say this is 👍 from me!

**Comment by nosolosw** [comment]
> I left a comment for something to fix.

@ItsJonQ I don't seem to find any comment in the PR?

**Review by ItsJonQ — CHANGES_REQUESTED** [review]
@nosolosw Oops! I forgot to hit Request changes! 🙇

File: lib/global-styles.php [hunk 1]
```diff
@@ -1,30 +1,37 @@
...
-		$separator     = $is_start ? '' : '--';
-		$new_key       = ( $prefix ? $prefix . $separator : '' ) . $processed_key;
+	foreach ( $global_styles as $key => $value ) {
+		$new_key = $prefix . str_replace( '/', '-', $key );
 
 		if ( is_array( $value ) ) {
 			$result = array_merge(
 				$result,
-				gutenberg_get_css_vars( $value, $new_key, false )
+				gutenberg_experimental_global_styles_get_css_vars( $value, $new_key . '-' )
```
[the remark this answers is not in the document]
Comment (nosolosw) [reply lib/global-styles.php]: Ah! But this was when we had different levels (globals vs blocks) so it was useful to differentiate the element part (global/block name) vs the CSS custom property name. Now that we don't have the element part (everything is prefixed by `--wp-`), is the `--` still useful?

[the remark this answers is not in the document]
Comment (nosolosw) [reply lib/global-styles.php]: I mean, now we have `--wp-color-primary`. Would we want to have `--wp--color-primary` instead?

[the remark this answers is not in the document]
Comment (nosolosw) [reply lib/global-styles.php]: Updated at https://github.com/WordPress/gutenberg/pull/20047/commits/7efa53e9ff73a472f4887aeb7ef763578eb72626 and https://github.com/WordPress/gutenberg/pull/20047/commits/798a0d3e4fd768689fd89175da8b5a824fd52c23

---

@jorgefilipecosta and I talked this over and I want to write down the trade-offs of the current approach so we have them documented somewhere until we find a better place.

Jorge's rationale for suggesting the double dash is this: in scoped styles (we store the CSS custom properties as block attributes) we need to deserialize the string into an object for editor things. In that case, we won't be able to distinguish between a valid CSS property name (ex: `line-height`) and a token we add (`background-color` as the resulting name of flattening the object `{background: { color: 'red' }}`). Let's work through an example. We've got this string we want to deserialize back into an object:

```
--wp-background-color: 'red';
--wp-line-height: 1.2;
```

How do we know which object to deserialize into without information about the object keys/CSS property names?

This is what we want:
```
{
    background: {
        color: 'red',
    },
    line-height: 1.2
}
```

A simple parser that creates one level per dash would return (which is not what we want, so we have to make the parser a bit more complex): 

```
{
    background: {
        color: 'red',
    },
    line {
        height: 1.2,
    }
}
```

By using double dashes the deserialization parser is simpler (each object level is separated by `--`):

```
--wp--background--color: 'red';
--wp--line-height: 1.2;
```

Personally, I'm not a fan of this implementation detail leaking into the API design/naming schema. However, I also think this is a fair approach to simplify the parser and get us going. The underlying assumption is that we store this information only once in the block attribute as a string, not as an object: when we cross that river, we'd have the opportunity to reevaluate this and change if necessary.

**Comment by nosolosw** [comment]
This PR is ready to land, although it only works in the block-editor. We're going to integrate Global Styles in the `edit-site` editor, for which we need to add support for some things in `edit-site`:

- hook admin_body_class: we use this to add the `wp-gs` class. Although the hook is fired, we need a way to identify the `edit-site` context server-side (something akin to `$current_screen->is_block_editor`).
- hook block settings: doesn't fire for `edit-site` context.
- hook enqueue_block_assets: doesn't fire for `edit-site` context.

I'm going to work on the `edit-site` things separately, so we can land this PR without being blocked.

**Comment by nosolosw** [comment]
I was able to make this work for `edit-site` in a decoupled and simple way, so I went ahead and added the code in this branch. This works now only with `edit-site` and frontend.

**Review by ItsJonQ — APPROVED** [review]
🚀 @nosolosw Awesome work!!!  Thank you for creating and iterating on this! Not only that, but you've also added it to FSE "Edit Site".

Testing locally, and it's working as outlined in your instructions.

From a code perspective, the mechanics of the PHP updates make sense of me.

**Comment by ItsJonQ** [comment]
@jorgefilipecosta If you're around, it would be great to get your thoughts on the PHP implementation 🙏 . Thank you!!

File: lib/experiments-page.php [hunk 2]
```diff
@@ -84,6 +84,17 @@ function gutenberg_initialize_experiments_settings() {
 			'id'    => 'gutenberg-full-site-editing-demo',
 		)
 	);
+	add_settings_field(
```
Comment (nosolosw) [inline lib/experiments-page.php]: This is unrelated to this PR, probably inadvertently added in a rebase. Will remove.

Comment (nosolosw) [reply lib/experiments-page.php]: Removed in https://github.com/WordPress/gutenberg/pull/20047/commits/6f1e87b1e041921dbc9e5a0582dd52920db71fbd

File: lib/global-styles.php [hunk 3]
```diff
@@ -34,98 +44,299 @@ function gutenberg_get_css_vars( $global_styles_branch, $prefix = '', $is_start
...
+ * @param array $post_status_filter Filter CPT by post status.
+ *                                  By default, only fetches published posts.
+ * @param bool  $should_create_draft Whether a new draft should be created
+ *                                   if no CPT was found. False by default.
+ * @return array Custom Post Type for the user's Global Styles.
+ */
+function gutenberg_experimental_global_styles_get_user_cpt( $post_status_filter = array( 'publish' ), $should_create_draft = false ) {
+	$user_cpt         = array();
+	$post_type_filter = 'wp_global_styles';
+	$post_name_filter = 'wp-global-styles-' . strtolower( wp_get_theme()->get( 'TextDomain' ) );
```
Comment (jorgefilipecosta) [inline lib/global-styles.php]: I think as an id of the theme we should use get_stylesheet() and not the 'TextDomain'. I think the database setting the WordPress core stores in the database to identify the current theme is the stylesheet.

File: lib/global-styles.php [hunk 4]
```diff
@@ -34,98 +44,299 @@ function gutenberg_get_css_vars( $global_styles_branch, $prefix = '', $is_start
...
-	if ( $theme_json_path ) {
-		$theme_json_global_styles = json_decode(
-			file_get_contents( $theme_json_path ),
+	if ( is_array( $recent_posts ) && ( count( $recent_posts ) === 1 ) ) {
+		$user_cpt = $recent_posts[0];
+	} elseif ( $should_create_draft ) {
+		$cpt_post_id = wp_insert_post(
+			array(
+				'post_content' => '{}',
+				'post_status'  => 'draft',
```
Comment (jorgefilipecosta) [inline lib/global-styles.php]: I think we may have a bug, in gutenberg_experimental_global_styles_get_user we are querying published posts, but we insert a draft post. On the next execution, a published post will still not exist and we will probably create another one.
I think we may be able to create the post with the published status ( the empty object will not do anything);

File: lib/global-styles.php [hunk 5]
```diff
@@ -34,98 +44,299 @@ function gutenberg_get_css_vars( $global_styles_branch, $prefix = '', $is_start
...
-	$inline_style .= '}';
 
 	wp_register_style( 'global-styles', false, array(), true, true );
 	wp_add_inline_style( 'global-styles', $inline_style );
 	wp_enqueue_style( 'global-styles' );
 }
-add_action( 'enqueue_block_assets', 'gutenberg_enqueue_global_styles_assets' );
 
 /**
- * Adds class wp-gs to the frontend body class if the theme defines a experimental-theme.json.
```
Comment (jorgefilipecosta) [inline lib/global-styles.php]: If we unconditionally add the wp-gs class we will break the styles of some blocks.
```
.wp-gs .wp-block-button__link:not(.has-background) {
	background-color: var(--wp--color--primary);
}
```

As the class is added even if the theme does not support global styles a theme that was changing the button color using the following code will stop working:
```
 .wp-block-button__link {
	background-color: red;
}
```

I think the whole idea of adding wp-gs was to mark themes that support global styles if we unconditionally add the class we would be able to not add it at all as well.
I know that for now we only add if the experiment is enabled but one day we remove the experiment and in that case, we would unconditionally add the class.
I think we should add the class when the experiment is enabled and when the theme supports global styles.

File: packages/block-library/src/button/style.scss [hunk 6]
```diff
@@ -37,7 +37,7 @@ $blocks-button__height: 56px;
 }
 
 .wp-gs .wp-block-button__link:not(.has-background) {
-	background-color: var(--wp-block-core-button--color--background, var(--wp-color--primary, $dark-gray-700));
+	background-color: var(--wp--color--primary);
```
Comment (jorgefilipecosta) [inline packages/block-library/src/button/style.scss]: Should we pass $dark-gray-700 as a fallback?

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @nosolosw nice work here 👍  generally the changes seem good and I think they are going in the right direction.

File: lib/global-styles.php (hunk 4 above)
Comment (nosolosw) [reply lib/global-styles.php]: Note the optional parameter `$should_create_draft`, which is false by default. We only insert a draft if there is none in the editor call `gutenberg_experimental_global_styles_get_user_cpt_id`. That's the only place we need a pre-existing CPT: the entities API only works with pre-existing objects, can't create new ones from the client-side.

File: lib/global-styles.php (hunk 5 above)
Comment (nosolosw) [reply lib/global-styles.php]: We already do that! The class is only added when the experiment is enabled, the theme has support, and the editor is site editor (for back-end, doesn't apply to front-end).

File: lib/global-styles.php (hunk 3 above)
Comment (nosolosw) [reply lib/global-styles.php]: The [template parts](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/template-part/index.php#L22) use the `TextDomain` for this, so I thought we may want to use the same mechanism for API consistency. I'm happy to change it to something else, but I guess it makes sense to update template parts as well?

File: packages/block-library/src/button/style.scss (hunk 6 above)
Comment (nosolosw) [reply packages/block-library/src/button/style.scss]: Note that `--wp--color--primary` is going to always have a value (user-defined, the theme default, or core default), so the fallback won't be necessary.
